src/chat-handler.ts

```diff
@@ -88,38 +88,42 @@ export class ChatHandler extends ChatModel {
     );
 
     this.updateWriters([{ username: 'AI' }]);
-    return this._aiProvider.chatModel
-      .invoke(messages)
-      .then(response => {
-        // Some providers may use text-completion models (for example ChromeAI),
-        // for which the response is a string.
-        const content = response.content ?? response;
-        const botMsg: IChatMessage = {
-          id: UUID.uuid4(),
-          body: content.toString(),
-          sender: { username: 'AI' },
-          time: Date.now(),
-          type: 'msg'
-        };
+
+    // create an empty message to be filled by the AI provider
+    const botMsg: IChatMessage = {
+      id: UUID.uuid4(),
+      body: '',
+      sender: { username: 'AI' },
+      time: Date.now(),
+      type: 'msg'
+    };
+
+    let content = '';
+
+    try {
+      for await (const chunk of await this._aiProvider.chatModel.stream(
+        messages
+      )) {
+        content += chunk.content;
+        botMsg.body = content;
         this.messageAdded(botMsg);
-        this._history.messages.push(botMsg);
-        return true;
-      })
-      .catch(reason => {
-        const error = getErrorMessage(this._aiProvider.name, reason);
-        const errorMsg: IChatMessage = {
-          id: UUID.uuid4(),
-          body: `**${error}**`,
-          sender: { username: 'ERROR' },
-          time: Date.now(),
-          type: 'msg'
-        };
-        this.messageAdded(errorMsg);
-        return false;
-      })
-      .finally(() => {
-        this.updateWriters([]);
-      });
+      }
+      this._history.messages.push(botMsg);
+      return true;
+    } catch (reason) {
+      const error = getErrorMessage(this._aiProvider.name, reason);
+      const errorMsg: IChatMessage = {
+        id: UUID.uuid4(),
+        body: `**${error}**`,
+        sender: { username: 'ERROR' },
+        time: Date.now(),
+        type: 'msg'
+      };
+      this.messageAdded(errorMsg);
```

Comment:
This seems to be doing the job, although there may be some future optimizations to be done in jupyter-chat.

Currently `jupyter-chat` seems to be removing the previous message before adding the one with the update here:

https://github.com/jupyterlab/jupyter-chat/blob/73512c16a6c602b81f873ecf9a3c20cb73cefd94/packages/jupyter-chat/src/model.ts#L435-L449

Maybe it should also compare the timestamps like the comment suggests?

Reply:
I think the comment do not refer to the correct line, the message is sent/updated above:
```
this.messageAdded(botMsg);
```

In your implementation, the timestamp is not updated, so the position of the message should not change. And it is probably better like this, to avoid confusion.
For comparison, streaming in jupyter-ai lead to [update_message](https://github.com/jupyterlab/jupyter-chat/blob/73512c16a6c602b81f873ecf9a3c20cb73cefd94/python/jupyterlab-chat/jupyterlab_chat/ychat.py#L142-L150) in the back-end, which do not update the timestamp.

We should probably homogenize this in `jupyter-chat`, and allow consumer extension to change or not the timestamp...

Reply:
Yes right it was about the line above (which Github shows as not modified).

> the timestamp is not updated, so the position of the message should not change. And it is probably better like this, to avoid confusion.

Yes, in terms of UX this seems fine.

Reply:
Also on another note for a different issue, after using the `jupyter-chat` API, I was wondering if changing to the following could make sense:

- `messageAdded` -> `addMessage`
- `messagesDeleted` -> `deleteMessage`

In general when working with extensions, when properties finish with `ed` they tend to refer to signals. Also it may more of a personal opinion but `addMessage` sounds more natural than `messageAdded` when interacting with the chat model.

Reply:
I agree, but in this case it could also be confusing with `sendMessage`  and `deleteMessage` that already exist.

This is the distinction between:
- the actions the user do in its chat (`sendMessage`  and `deleteMessage`) that must be propagated to the other users
- the actions done by remote users (`messageAdded` and `messagesDeleted`) and that must update the chat view...

In the case of `jupyterlite` chats, it is most confusing because there is no back-end, the `sendMessage` propagates itself the messages from the user and the AI.

It may worth an issue in `jupyter-chat` to discuss the naming.

Reply:
Yeah happy to open an issue on https://github.com/jupyterlab/jupyter-chat to discuss this 👍